include/boost/math/tools/promotion.hpp

```diff
@@ -101,51 +91,51 @@ namespace boost
+      template <> struct promote_args_2<double, double>{ using type = double; };
+      template <> struct promote_args_2<long double, long double> { using type = long double; };
+      template <> struct promote_args_2<int, int> {  using type = double; };
+      template <> struct promote_args_2<int, float> {  using type = double; };
+      template <> struct promote_args_2<float, int> {  using type = double; };
+      template <> struct promote_args_2<int, double> {  using type = double; };
+      template <> struct promote_args_2<double, int> {  using type = double; };
+      template <> struct promote_args_2<int, long double> {  using type = long double; };
+      template <> struct promote_args_2<long double, int> {  using type = long double; };
+      template <> struct promote_args_2<float, double> {  using type = double; };
+      template <> struct promote_args_2<double, float> {  using type = double; };
+      template <> struct promote_args_2<float, long double> {  using type = long double; };
+      template <> struct promote_args_2<long double, float> {  using type = long double; };
+      template <> struct promote_args_2<double, long double> {  using type = long double; };
+      template <> struct promote_args_2<long double, double> {  using type = long double; };
 
       template <class T1, class T2=float, class T3=float, class T4=float, class T5=float, class T6=float>
       struct promote_args
       {
-         typedef typename promote_args_2<
-            typename remove_cv<T1>::type,
+         using type = typename promote_args_2<
+            typename std::remove_cv<T1>::type,
             typename promote_args_2<
-               typename remove_cv<T2>::type,
+               typename std::remove_cv<T2>::type,
                typename promote_args_2<
-                  typename remove_cv<T3>::type,
+                  typename std::remove_cv<T3>::type,
                   typename promote_args_2<
-                     typename remove_cv<T4>::type,
+                     typename std::remove_cv<T4>::type,
                      typename promote_args_2<
-                        typename remove_cv<T5>::type, typename remove_cv<T6>::type
+                        typename std::remove_cv<T5>::type, typename std::remove_cv<T6>::type
                      >::type
                   >::type
                >::type
             >::type
-         >::type type;
+         >::type;
 
 #ifdef BOOST_MATH_NO_LONG_DOUBLE_MATH_FUNCTIONS
          //
          // Guard against use of long double if it's not supported:
          //
-         BOOST_STATIC_ASSERT_MSG((0 == ::boost::is_same<type, long double>::value), "Sorry, but this platform does not have sufficient long double support for the special functions to be reliably implemented.");
+         BOOST_STATIC_ASSERT_MSG((0 == std::is_same<type, long double>::value), "Sorry, but this platform does not have sufficient long double support for the special functions to be reliably implemented.");
```

Comment:
These BOOST_STATIC_ASSERT's could be real `static_assert`'s now, but we can do that later ;)